### PR TITLE
[Core] Adding RemoveOrphanNodesFromSubModelParts

### PR DESCRIPTION
**Pull request by loumalouomega**
**📝 Description**

Part of https://github.com/KratosMultiphysics/Kratos/pull/10563

I isolate the changes without modifying behaviour, so we can merge this faster. Test added

**🆕 Changelog**

- [Adding RemoveOrphanNodesFromSubModelParts](https://github.com/KratosMultiphysics/Kratos/commit/5e35f581dc1c315837cec550c979c77445935b29)
- [Improve code](https://github.com/KratosMultiphysics/Kratos/commit/b72d66fd7ce8362a590e1ccbc452cbbccef6e73e)
- [Adding test](https://github.com/KratosMultiphysics/Kratos/commit/989b384be613eddc59cf8fa0435185b2114e5230)

### CODE REVIEW

File: kratos/utilities/auxiliar_model_part_utilities.cpp [hunk 1]
```diff
@@ -376,6 +376,38 @@ void AuxiliarModelPartUtilities::RemoveConditionsAndBelongingsFromAllLevels(cons
...
 /***********************************************************************************/
 
+void AuxiliarModelPartUtilities::RemoveNodesFromSubModePartsWithoutCorrespondingEntities()
+{
+    VariableUtils variable_utils;
+    for (auto& r_sub_model_part : mrModelPart.SubModelParts()) {
+        if (r_sub_model_part.NumberOfNodes() > 0) {
+            auto& r_array_nodes = r_sub_model_part.Nodes();
+            variable_utils.ResetFlag(TO_ERASE, r_array_nodes);
+            variable_utils.SetFlag(TO_ERASE, true, r_array_nodes);
```
Comment (rubenzorrilla) [inline kratos/utilities/auxiliar_model_part_utilities.cpp]: Why is the reset required if you're setting all to `true`?

File: kratos/utilities/auxiliar_model_part_utilities.cpp [hunk 2]
```diff
@@ -376,6 +376,38 @@ void AuxiliarModelPartUtilities::RemoveConditionsAndBelongingsFromAllLevels(cons
 /***********************************************************************************/
 /***********************************************************************************/
 
+void AuxiliarModelPartUtilities::RemoveNodesFromSubModePartsWithoutCorrespondingEntities()
```
Comment (rubenzorrilla) [inline kratos/utilities/auxiliar_model_part_utilities.cpp]: The name is way misleading. Without checking the implementation what comes to my mind is that you want to remove the nodes but keeping the conditions and elements. Besides there is a typo. I'd suggest calling it `RemoveOrphanNodesFromSubModelParts`.

File: kratos/utilities/auxiliar_model_part_utilities.cpp [hunk 3]
```diff
@@ -376,6 +376,38 @@ void AuxiliarModelPartUtilities::RemoveConditionsAndBelongingsFromAllLevels(cons
...
+                    auto& r_geometry = r_cond.GetGeometry();
+                    for (auto& r_node : r_geometry) {
+                        r_node.Set(TO_ERASE, false);
+                    }
+                }
+            }
+            r_sub_model_part.RemoveNodes(TO_ERASE);
+        }
+    }
+    variable_utils.ResetFlag(TO_ERASE, mrModelPart.Nodes());
```
Comment (rubenzorrilla) [inline kratos/utilities/auxiliar_model_part_utilities.cpp]: This shouldn't be required as all the `TO_ERASE` nodes must have been removed.
```suggestion
```

**Comment by loumalouomega** [comment]
Done @rubenzorrilla

File: kratos/python/add_other_utilities_to_python.cpp [hunk 4]
```diff
@@ -410,6 +410,7 @@ void AddOtherUtilitiesToPython(pybind11::module &m)
         .def("RemoveConditionAndBelongingsFromAllLevels", [](AuxiliarModelPartUtilities& rAuxiliarModelPartUtilities, ModelPart::ConditionType::Pointer pThisCondition, Flags IdentifierFlag) { rAuxiliarModelPartUtilities.RemoveConditionAndBelongingsFromAllLevels(pThisCondition, IdentifierFlag);})
         .def("RemoveConditionAndBelongingsFromAllLevels", [](AuxiliarModelPartUtilities& rAuxiliarModelPartUtilities, ModelPart::ConditionType::Pointer pThisCondition, Flags IdentifierFlag, ModelPart::IndexType ThisIndex) { rAuxiliarModelPartUtilities.RemoveConditionAndBelongingsFromAllLevels(pThisCondition, IdentifierFlag, ThisIndex);})
         .def("RemoveConditionsAndBelongingsFromAllLevels", &Kratos::AuxiliarModelPartUtilities::RemoveConditionsAndBelongingsFromAllLevels)
+        .def("RemoveNodesFromSubModePartsWithoutCorrespondingEntities", &Kratos::AuxiliarModelPartUtilities::RemoveNodesFromSubModePartsWithoutCorrespondingEntities)
```
Comment (philbucher) [inline kratos/python/add_other_utilities_to_python.cpp]:
```suggestion
        .def("RemoveOrphanNodesFromSubModelParts", &Kratos::AuxiliarModelPartUtilities::RemoveOrphanNodesFromSubModelParts)
```

File: kratos/utilities/auxiliar_model_part_utilities.cpp (hunk 2 above)
Comment (loumalouomega) [reply kratos/utilities/auxiliar_model_part_utilities.cpp]: Done

Comment (loumalouomega) [reply kratos/utilities/auxiliar_model_part_utilities.cpp]: I  don't why I didn't see the notification from this message...

File: kratos/python/add_other_utilities_to_python.cpp (hunk 4 above)
Comment (loumalouomega) [reply kratos/python/add_other_utilities_to_python.cpp]: Done

File: kratos/utilities/auxiliar_model_part_utilities.cpp [hunk 5]
```diff
@@ -376,6 +376,37 @@ void AuxiliarModelPartUtilities::RemoveConditionsAndBelongingsFromAllLevels(cons
...
+void AuxiliarModelPartUtilities::RemoveOrphanNodesFromSubModelParts()
+{
+    VariableUtils variable_utils;
+    for (auto& r_sub_model_part : mrModelPart.SubModelParts()) {
+        if (r_sub_model_part.NumberOfNodes() > 0) {
+            auto& r_array_nodes = r_sub_model_part.Nodes();
+            variable_utils.ResetFlag(TO_ERASE, r_array_nodes);
+            variable_utils.SetFlag(TO_ERASE, true, r_array_nodes);
+            // Check orphans nodes
+            if (r_sub_model_part.NumberOfElements() > 0 || r_sub_model_part.NumberOfConditions() > 0) {
```
Comment (philbucher) [inline kratos/utilities/auxiliar_model_part_utilities.cpp]: what about geometries? :thinking:

Comment (loumalouomega) [reply kratos/utilities/auxiliar_model_part_utilities.cpp]: image

Comment (loumalouomega) [reply kratos/utilities/auxiliar_model_part_utilities.cpp]: Done

**Review by rubenzorrilla — COMMENTED** [review]
Nice, but can we also check the geometry stuff in the test you added?

**Review by rubenzorrilla — DISMISSED** [review]
Fine for me if CI passes. Thanks!!

**Comment by loumalouomega** [comment]
> Fine for me if CI passes. Thanks!!

I needed to fix a typo in the test, sorry

File: kratos/utilities/auxiliar_model_part_utilities.cpp (hunk 1 above)
Comment (rubenzorrilla) [reply kratos/utilities/auxiliar_model_part_utilities.cpp]: Sorry, this is still pending.

File: kratos/utilities/auxiliar_model_part_utilities.cpp [hunk 6]
```diff
@@ -376,6 +376,44 @@ void AuxiliarModelPartUtilities::RemoveConditionsAndBelongingsFromAllLevels(cons
 /***********************************************************************************/
 /***********************************************************************************/
 
+void AuxiliarModelPartUtilities::RemoveOrphanNodesFromSubModelParts()
+{
+    VariableUtils variable_utils;
+    for (auto& r_sub_model_part : mrModelPart.SubModelParts()) {
+        if (r_sub_model_part.NumberOfNodes() > 0) {
+            auto& r_array_nodes = r_sub_model_part.Nodes();
+            variable_utils.ResetFlag(TO_ERASE, r_array_nodes);
```
Comment (rubenzorrilla) [inline kratos/utilities/auxiliar_model_part_utilities.cpp]:
```suggestion
```
I think that this should be enough.

**Review by rubenzorrilla — APPROVED** [review]
Now should be fine.

**Pull request merged**